### PR TITLE
mgr/cephadm: Add Cert/Key Store 

### PR DESCRIPTION
**Pull request by adk3798**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/tests/test_cephadm.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/tests/test_cephadm.py:1742]: yeah, the current output is a bit ugly. I wanted the option to show all the cert types just for people to be able to know all the different ones we might store here, but it's probably better to normally just list the present ones and then hide the name of all the possible ones behind a flag or a different command.

File: src/pybind/mgr/cephadm/inventory.py
Comment (rkachach) [inline src/pybind/mgr/cephadm/inventory.py:1968]: Plz, can you explain with some code comment what are we trying to achieve here? I find the logic a little bit confusing.

File: src/pybind/mgr/cephadm/inventory.py
Comment (rkachach) [inline src/pybind/mgr/cephadm/inventory.py:1770]: Do we really need this field? we can inspect the cert to see if it was self-generated or not? is it to make this check faster?

File: src/pybind/mgr/cephadm/inventory.py
Comment (rkachach) [inline src/pybind/mgr/cephadm/inventory.py:2029]: same comment as before... just to see if we can simplify this logic

File: src/pybind/mgr/cephadm/inventory.py
Comment (rkachach) [inline src/pybind/mgr/cephadm/inventory.py:1872]: Once we move the info of mappings to the corresponding we shouldn't need this maps? should we?

File: src/pybind/mgr/cephadm/tests/test_cephadm.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/tests/test_cephadm.py:1742]: I think we can provide some options to tweak this command so users i.e can:

- show all the certificates
- show details about certificates (due date, owner, etc)
- filter certificates by some field (service, etc)
- ..

I'm not asking to change the current PR but just giving some more ideas about options that could be useful for the users.

**Review by rkachach — COMMENTED** [review]
As we discussed on the daily we should move all the "specific" code that handles certs/keys loading/saving. All the code has to be generic and we should move any specific code to the corresponding spec classes. As we agreed, I'm OK with addressing this on a separate PR.

File: src/pybind/mgr/cephadm/inventory.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py:1869]: You can't move all of these to the spec classes. The ingress, rgw, and iscsi related ones can, but prometheus, grafana, etc. were never part of a spec. We just generate those certs at misc. spots in the codebase and then were either storing them in the config-key store or just not storing them at all.

File: src/pybind/mgr/cephadm/inventory.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py:1904]: same issue as described in https://github.com/ceph/ceph/pull/55564#discussion_r1644650076

File: src/pybind/mgr/cephadm/inventory.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py:1845]: If we do that it should part of a full refactor of inventory.py. Right now all of our classes meant to act as interfaces between in memory storage and the config-key store (HostCache, SpecStore, etc.) live in this file. Having a `inventory` subdir in pybind/mgr/cephadm/ that has separate files for the various store classes makes sense, but I don't like just moving this single one now.

File: src/pybind/mgr/cephadm/inventory.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py:1968]: The outer loop goes through all the potential certs/keys we the cert store is aware of. `self.service_name_cert` is the list of these certs that is stored as a mapping from service names to certs. `self.host_cert` is the list of all certs that are stored as mappings from hosts to certs. Since those are both one level deep strings in the dict, you can handle the identically, hence the single loop going through the keys in `self.known_certs`. Beyond that, we are building this `ls` dict that basically just checks for each cert we could have if there is actually one there. Then `tmp` is just for tracking within a single type of cert (e.g. rgw_frontend_ssl_cert or ingress' ssl_cert) what we have for that cert. It then checks if that tmp dict actually is populated or not. If it's not (`if len(tmp) == 0`) then it just sets that key to false, and otherwise sets it to the tmp dict we just built. PArt of the reason it's so confusing is the structure it's trying to build is not intuitive. 
```
{
    "rgw_frontend_ssl_cert": false,
    "iscsi_ssl_cert": false,
    "nvmeof_client_cert": false,
    "nvmeof_server_cert": false,
    "ingress_ssl_cert": false,
    "agent_endpoint_root_cert": false,
    "service_discovery_root_cert": false,
    "grafana_cert": false,
    "alertmanager_cert": {
        "host1": true,
        "host2": true
    },
    "prometheus_cert": false,
    "node_exporter_cert": false
}
```
I will probably actually change this output later as discussed in https://github.com/ceph/ceph/pull/55564#discussion_r1643147887

File: src/pybind/mgr/cephadm/inventory.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py:1770]: Since we can tell when the cert is being stored if it's user made or not and it's inexpensive to store that info, I thought we might as well try to track it. I like having it tracked explicitly rather than trying to figure it out whenever we may need it. That's sort of a general theme of how this cert store is being built with the types of certs we store being explicitly listed.

File: src/pybind/mgr/cephadm/inventory.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py:2029]: same response https://github.com/ceph/ceph/pull/55564#discussion_r1644663508

File: src/pybind/mgr/cephadm/inventory.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/inventory.py:1869]: @adk3798 I see your point .. monitoring has always been treated differently which is a problem. Can't we use the class `class MonitoringSpec` to hold this info?

File: src/pybind/mgr/cephadm/inventory.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/inventory.py:1845]: I see, if there're plans to refactor the inventory then I'm OK with keeping it there and moving it later :+1:

File: src/pybind/mgr/cephadm/inventory.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/inventory.py:1968]: @adk3798 thank you very much for the detailed explanation. I think the code may be simplified to something as bellow (bool operator can help to evaluate dictionaries):

```
def cert_ls(self) -> Dict[str, Union[bool, Dict[str, bool]]]:
    ls = {}
    for k, v in self.known_certs.items():
        if k in self.service_name_cert or k in self.host_cert:
            tmp = {key: True for key in v if v[key]}
            ls[k] = tmp if tmp else False
        else:
            ls[k] = bool(v)
    return ls
```

File: src/pybind/mgr/cephadm/inventory.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/inventory.py:1770]: I see, I'm OK with keeping the `user_made` boolean :+1:

File: src/pybind/mgr/cephadm/inventory.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py:1869]: in theory, yes, but that would be a pretty big configuration change to move those certs into the spec. It might go okay, although I think we'd need to have a spec for each monitoring daemon type (if they don't already exist. I think most, if not all, do). I'm open to the idea. Maybe a topic for the weekly.

File: src/pybind/mgr/cephadm/inventory.py
Comment (adk3798) [reply src/pybind/mgr/cephadm/inventory.py]: updated with this logic

File: src/pybind/mgr/cephadm/inventory.py
Comment (rkachach) [inline src/pybind/mgr/cephadm/inventory.py:1801]: this is equivalent to: user_made = user_made.lower() == 'true'

File: src/pybind/mgr/cephadm/inventory.py
Comment (rkachach) [inline src/pybind/mgr/cephadm/inventory.py:1843]: same comment as above

File: src/pybind/mgr/orchestrator/module.py
Comment (rkachach) [inline src/pybind/mgr/orchestrator/module.py:1156]: what about something like?

**Note**: I didn't test the code just as an idea on how can we use resistivity to introduce multi-level indentation. 

```
def process_dict(d, level=0):
    result_str = ""
    indent = "  " * level
    for k, v in d.items():
        if isinstance(v, dict):
            result_str += f'{indent}{k}\n'
            result_str += process_dict(v, level + 1)
        else:
            result_str += f'{indent}{k} - {v}\n'
    return result_str
```

**Review by rkachach — COMMENTED** [review]
LGTM in general, just some more few minor comments. Overall I think the code should be good to merge if testing is passing. Additional PR would be needed to address the rest of the comments.

File: src/pybind/mgr/orchestrator/module.py
Comment (rkachach) [inline src/pybind/mgr/orchestrator/module.py:1156]: what about something like?

**Note**: I didn't test the code just as an idea on how can we use resistivity to introduce multi-level indentation. 

```
def process_dict(d, level=0):
    result_str = ""
    indent = "  " * level
    for k, v in d.items():
        if isinstance(v, dict):
            result_str += f'{indent}{k}\n'
            result_str += process_dict(v, level + 1)
        else:
            result_str += f'{indent}{k} - {v}\n'
    return result_str
```

**Review by rkachach — APPROVED** [review]
LGTM. Minor comments can be addressed on follow-up PR.

**Comment by rkachach** [comment]
jenkins test make check

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-06-28_04:58:18-orch:cephadm-wip-adk-testing-2024-06-27-0902-distro-default-smithi/

Failures all fell under

- cluster log failures (known issue, being cleaned up)
- mds_upgrade_Sequence (known issue)
- staggered upgrade with agent on (known issue)

Nothing to block merging

**Comment by adk3798** [comment]
jenkins test make check

**Pull request merged**
